File: src/World/Objects/Leaves.cpp
/*
   Copyright 2010-2011 Tyslenko Max, Bogatirev Pavel.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "World/Objects/Leaves.hpp"

Leaves::Leaves(uint16_t maxLeaves) : mFirstTime(true)
{
    srand(time(NULL));
    mGC = appManager().getGraphic();
    mMaxObjects = maxLeaves;

    functionValue = +1;
}

void Leaves::init()
{
    float windPower = worldManager().getWindPower();
    for (uint16_t i=0; i < mMaxObjects; ++i)
    {
        mLeaves.push_back(LeafData());
        processLeaves(mGC, windPower, mLeaves[i], true);
    }
}

void Leaves::processLeaves(CL_GraphicContext &gc, float windPower, LeafData &current, bool firstTime)
{
    current.timer = current.addedAngle = 0;
    current.x_speed = current.y_speed = 0;

    current.animation = false;

    // The height where new leaf will be located
    current.y = ScreenResolutionY - ScreenResolutionY*((rand()%6+2)/10.0);

    // Some magic numbers for the leaf trajectory
    current.k1 = rand() % 30 + 20;
    current.k2 = (rand() % 7 + 3) / 10.0;

    // Load some random leaf surface
    uint16_t size = resourceManager().sectionHandle("Leaves").get_child_nodes().get_length();
    current.leafType = rand() % size;

    current.imageHandle = resourceManager().getSprite("Leaves", boost::lexical_cast<std::string>(current.leafType));

    // Rotate leaf surface at some random angle
    uint16_t angle = rand() % 360 + 1;
    current.imageHandle.set_angle(CL_Angle::from_degrees(angle));

    float scale = static_cast<float>(rand()% 55 + 50) / 100.0f;
    current.imageHandle.set_scale(scale, scale);

    current.imageHandle.set_color(CL_Colorf(17/255.0f, 30/255.0f, 35/255.0f));

    current.speed_koef = rand()%70 + 50;

    // Where to create new leaf (refers to the wind blowing direction)
    if (firstTime)
    {
        current.x = rand() % ScreenResolutionX;
        current.timeout = 0;
    }
    else
    {
        if (windPower < 0)
            current.x = ScreenResolutionX;
        else
            current.x = 0 - current.imageHandle.get_width();

        current.timeout = rand() % 250;
    }
}

void Leaves::update(float elapsed)
{

}

void Leaves::updateVisual(float newX, float newY)
{
    float elapsed = appManager().getElapsed() / 1000.0f;
    float windPower = worldManager().getWindPower();
    float newSpeed1 = windPower * elapsed * 0.4;
    float newSpeed2 = Gravity * elapsed;
    for (uint16_t i=0; i < mMaxObjects; i++)
    {
        LeafData &current = mLeaves[i];
        if (current.timeout > 0)
        {
            current.timeout--;
        }
        else
        {
            if (windPower > 0) {
                if (current.x > ScreenResolutionX + 10 || current.y > ScreenResolutionY)
                    processLeaves(mGC, windPower, current);
            } else {
                if (current.x < -current.imageHandle.get_width() || current.y > ScreenResolutionY)
                    processLeaves(mGC, windPower, current);
            }

            current.x += current.x_speed * elapsed;
            current.y += current.y_speed * elapsed;

            float y = current.k1 * sinf(current.k2 * Deg2Rad(current.x)) + current.y;

            float tempValue = current.k1 * current.k2 * cosf(current.k2 * Deg2Rad(current.x));

            if (tempValue * functionValue < 0)
            {
                current.animation = true;
            }
            functionValue = tempValue;

            if (current.animation)
            {
                current.timer += elapsed * 1000.0;
                if (current.timer >= 100)
                {
                    uint16_t angle = rand() % 15 + 5;
                    current.addedAngle += angle;

                    current.imageHandle.rotate_pitch(CL_Angle::from_degrees(angle));

                    current.timer = 0;

                    if (current.addedAngle >= 180)
                    {
                        current.animation = false;
                        current.addedAngle = 0;
                    }
                }
            }

            current.x_speed = current.speed_koef * newSpeed1;
            current.y_speed = newSpeed2;
            current.imageHandle.draw(mGC, current.x, y);
        }
    }
}

boost::shared_ptr<Object> Leaves::ParseLeaves(CL_DomElement tag)
{
    float max = 0;
    if(tag.has_attribute("maxLeaves"))
    {
        std::string maxStr = tag.get_attribute("maxLeaves").c_str();
        max = boost::lexical_cast<float>(maxStr);
    }

    float area = Pixels2Meters(ScreenResolutionX) * Pixels2Meters(ScreenResolutionY);
    return boost::shared_ptr<Object>(new Leaves(max * area));
}

bool Leaves::checkCollision(CL_Pointf point) { return false; }

CL_Rectf Leaves::getRectangle()
{
    return CL_Rectf(0, 0, ScreenResolutionX, ScreenResolutionY);
}

CL_Pointf Leaves::getPosition()
{
    return CL_Pointf(0, 0);
}

void Leaves::setPosition(CL_Pointf newPos)
{
}

File: bin/media/objects/Leaves.py
import Core.LogManager
import Core.ApplicationManager
import Core.WorldManager
import World.Objects.TypesManager
from World.Objects.Object import *
import Core.Utils
import random

class LeafData:
    def __init__(self):
        self.x = self.y = self.x_speed = self.y_speed = 0.0
        self.k1 = self.k2 = self.speed_koef = 0.0
        self.addedAngle = 0.0
        self.y_sin = 0

        self.image = None
        self.animation = False

class Leaves(Object):
    def __init__(self, maxObjects):
        super(Leaves, self).__init__()

        self.mLeaves = []

        # Precalculated value
        self.mMaxObjects = int(maxObjects)
        self.functionValue = +1

        self.mAppManagerHandle = Core.ApplicationManager.getInstance()
        self.mGC = self.mAppManagerHandle.GetGraphic()
        self.mWorldManagerHandle = Core.WorldManager.getInstance()

    def processLeaves(self, windPower, curObject, firsTime):
        curObject.x_speed = curObject.y_speed = 0.0
        curObject.timer = curObject.addedAngle = 0.0
        curObject.animation = False

        curObject.y = Core.Utils.ScreenResolutionY - Core.Utils.ScreenResolutionY * random.uniform(0.2, 0.8)

        curObject.k1 = random.randint(20, 50)
        curObject.k2 = random.uniform(0.3, 1.0)

        size = Core.ResourceManager.SectionNumChildren("Leaves")
        leaf = random.randint(0, size-1)
        curObject.image = Core.ResourceManager.getInstance().GetSprite("Leaves", str(leaf))
        angle = random.randint(1, 360)
        curObject.image.SetAngle(CL_Angle.FromDegrees(angle))

        scale = random.uniform(0.5, 0.9)
        curObject.image.SetScale(scale, scale)

        color = ColorRGB(17, 30, 35)
        curObject.image.SetColor(color)

        curObject.speed_koef = random.randint(50, 120)

        if not firsTime:
            if windPower < 0:
                curObject.x = Core.Utils.ScreenResolutionX + curObject.image.GetWidth()
            else:
                curObject.x = 0 - curObject.image.GetWidth()

            curObject.timeout = random.randint(0, 250)
        else:
            curObject.x = random.randint(0, Core.Utils.ScreenResolutionX)
            curObject.timeout = 0

    def Init(self):
        windPower = self.mWorldManagerHandle.GetWindPower()

        for i in xrange(self.mMaxObjects-1):
            self.mLeaves.append(LeafData())
            self.processLeaves(windPower, self.mLeaves[i], True)

    def Update(self, elapsed):
        windPower = self.mWorldManagerHandle.GetWindPower()

        elapsed_ = elapsed # /  1000
        newSpeed1 = windPower * elapsed_ * 0.4
        newSpeed2 = Core.Utils.Gravity * elapsed_

        for i in xrange(self.mMaxObjects-1):
            current = self.mLeaves[i]
            if current.timeout > 0:
                current.timeout -= 1
            else:
                if windPower > 0:
                    if current.x > Core.Utils.ScreenResolutionX + current.image.GetWidth():
                        self.processLeaves(windPower, current, False)
                else:
                    if current.x < -current.image.GetWidth():
                        self.processLeaves(windPower, current, False)

                current.x += current.x_speed * elapsed_
                current.y += current.y_speed * elapsed_

                # Process animation
                y = current.k1 * math.sin(current.k2 * Core.Utils.Deg2Rad(current.x)) + current.y
                tempValue = current.k1 * current.k2 * math.cos(current.k2 * Core.Utils.Deg2Rad(current.x))
                current.y_sin = y

                if tempValue * self.functionValue < 0:
                    current.animation = True
                functionValue = tempValue

                if current.animation:
                    angle = random.randint(1, 3)
                    current.addedAngle += angle
                    current.image.RotatePitch(CL_Angle.FromDegrees(angle))

                    if current.addedAngle >= 180:
                        current.animation = False
                        current.addedAngle = 0

                current.x_speed = current.speed_koef * newSpeed1
                current.y_speed = newSpeed2

    def UpdateVisual(self, newX, newY):
        for i in xrange(self.mMaxObjects-1):
            current = self.mLeaves[i]
            if current.timeout <= 0:
                current.image.Draw(self.mGC, current.x, current.y_sin)

    def GetPosition(self):
        return CL_Pointf(0, 0)

    def SetPosition(self):
        print 'Setting position'

    def GetRectangle(self):
        return CL_Rectf(0, 0, Core.Utils.ScreenResolutionX, Core.Utils.ScreenResolutionY)

    def CheckCollision(self):
        return False

    @staticmethod
    def ParseLeaves(node):
        maxLeaves = 0

        if node.HasAttribute("maxLeaves"):
            maxLeaves = float(node.GetAttribute("maxLeaves"))

        area = Core.Utils.Pixels2Meters(Core.Utils.ScreenResolutionX) * \
              Core.Utils.Pixels2Meters(Core.Utils.ScreenResolutionY)

        return Leaves(area * maxLeaves)

World.Objects.TypesManager.RegisterParser("LeavesObject", Leaves.ParseLeaves)
